turbopack/crates/turbo-tasks-backend/src/backend/storage.rs

```diff
@@ -0,0 +1,390 @@
+where
+    T::Key: Indexed,
+    T::Value: PartialEq,
+{
+    pub fn has(&self, item: &mut T) -> bool {
+        let (key, value) = take(item).into_key_and_value();
+        let result = if let Some(stored_value) = self.get(&key) {
+            *stored_value == value
+        } else {
+            false
+        };
+        *item = T::from_key_and_value(key, value);
+        result
+    }
+}
+
+pub struct Storage<K, T: KeyValuePair>
+where
+    T::Key: Indexed,
+{
+    map: DashMap<K, InnerStorage<T>, BuildHasherDefault<FxHasher>>,
+}
+
+impl<K, T> Storage<K, T>
+where
+    T: KeyValuePair,
+    T::Key: Indexed,
+    K: Eq + std::hash::Hash + Clone,
+{
+    pub fn new() -> Self {
+        let shard_amount =
+            (available_parallelism().map_or(4, |v| v.get()) * 64).next_power_of_two();
+        Self {
+            map: DashMap::with_capacity_and_hasher_and_shard_amount(
+                1024 * 1024,
+                Default::default(),
+                shard_amount,
+            ),
+        }
+    }
+
+    pub fn access_mut(&self, key: K) -> StorageWriteGuard<'_, K, T> {
+        let inner = match self.map.entry(key) {
+            dashmap::mapref::entry::Entry::Occupied(e) => e.into_ref(),
+            dashmap::mapref::entry::Entry::Vacant(e) => e.insert(InnerStorage::new()),
+        };
+        StorageWriteGuard {
+            inner: inner.into(),
+        }
+    }
+
+    pub fn access_pair_mut(
+        &self,
+        key1: K,
+        key2: K,
+    ) -> (StorageWriteGuard<'_, K, T>, StorageWriteGuard<'_, K, T>) {
+        let (a, b) = get_multiple_mut(&self.map, key1, key2, || InnerStorage::new());
+        (
+            StorageWriteGuard { inner: a },
+            StorageWriteGuard { inner: b },
+        )
+    }
+}
+
+pub struct StorageWriteGuard<'a, K, T>
+where
+    T: KeyValuePair,
+    T::Key: Indexed,
+{
+    inner: RefMut<'a, K, InnerStorage<T>, BuildHasherDefault<FxHasher>>,
+}
+
+impl<'a, K, T> Deref for StorageWriteGuard<'a, K, T>
+where
+    T: KeyValuePair,
+    T::Key: Indexed,
+    K: Eq + Hash,
+{
+    type Target = InnerStorage<T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<'a, K, T> DerefMut for StorageWriteGuard<'a, K, T>
+where
+    T: KeyValuePair,
+    T::Key: Indexed,
+    K: Eq + Hash,
+{
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+#[macro_export]
+macro_rules! get {
+    ($task:ident, $key:ident $input:tt) => {
+        if let Some($crate::data::CachedDataItemValue::$key { value }) = $task.get(&$crate::data::CachedDataItemKey::$key $input).as_ref() {
+            Some(value)
+        } else {
+            None
+        }
+    };
+    ($task:ident, $key:ident) => {
+        if let Some($crate::data::CachedDataItemValue::$key { value }) = $task.get(&$crate::data::CachedDataItemKey::$key {}).as_ref() {
+            Some(value)
+        } else {
+            None
+        }
+    };
```

Comment:
I think the two macro branches can be combined:

- Use `$(...)?` for a zero-or-one repetition: https://veykril.github.io/tlborm/decl-macros/macros-methodical.html#repetitions
- I don't think the `{}` curly braces are required when the enum variant has no fields.

If you can't combine the branches, best-practice is to make the macro recursive so that the branch without `$input:tt` calls the macro again with a default argument (`{}`).

I think you can shorten/DRY a lot of the other macros here with similar tricks.

```suggestion
    ($task:ident, $key:ident $($input:tt)?) => {
        if let Some($crate::data::CachedDataItemValue::$key { value }) = $task.get(&$crate::data::CachedDataItemKey::$key $($input)?).as_ref() {
            Some(value)
        } else {
            None
        }
    };
```

Reply:
`{}` is required, but I make it recursive